File: macos/Flash/Data/FlashDatabase.swift
// Copyright (c) 2026 Fengyuan Jiang
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

import Foundation
import SwiftData

enum FlashDatabase {
    /// 容器创建结果。didFallbackToMemory 为 true 表示持久化容器创建失败、
    /// 已降级为内存容器（App 可用但数据不持久），调用方可据此提示用户。
    struct ContainerResult {
        let container: ModelContainer
        let didFallbackToMemory: Bool
    }

    /// 持久化容器创建失败时降级为内存容器，不再 fatalError 导致启动即崩。
    static func makeContainer(inMemory: Bool = false) -> ModelContainer {
        makeContainerWithFallback(inMemory: inMemory).container
    }

    static func makeContainerWithFallback(inMemory: Bool = false) -> ContainerResult {
        let schema = Schema([LogEntity.self, EmotionEntity.self, TaskEntity.self])
        let config = ModelConfiguration(schema: schema, isStoredInMemoryOnly: inMemory)
        do {
            return ContainerResult(container: try ModelContainer(for: schema, configurations: [config]),
                                   didFallbackToMemory: false)
        } catch {
            print("SwiftData 容器创建失败，尝试降级为内存容器: \(error)")
            let memoryConfig = ModelConfiguration(schema: schema, isStoredInMemoryOnly: true)
            if let fallback = try? ModelContainer(for: schema, configurations: [memoryConfig]) {
                print("已降级为内存容器：本次运行数据不会持久化")
                return ContainerResult(container: fallback, didFallbackToMemory: true)
            }
            fatalError("无法创建 SwiftData 容器（内存降级亦失败）: \(error)")
        }
    }
}

File: android/app/src/main/java/com/flash/app/data/db/FlashDatabase.kt
// Copyright (c) 2026 Fengyuan Jiang
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

package com.flash.app.data.db

import androidx.room.Database
import androidx.room.RoomDatabase
import androidx.room.migration.Migration
import androidx.sqlite.db.SupportSQLiteDatabase

@Database(
    entities = [LogEntity::class, EmotionEntity::class, IdeaViewStateEntity::class, TaskEntity::class],
    version = 4,
    exportSchema = true,
)
abstract class FlashDatabase : RoomDatabase() {
    abstract fun logDao(): LogDao
    abstract fun emotionDao(): EmotionDao
    abstract fun taskDao(): TaskDao

    companion object {
        /** v2 仅增加本地 Idea 阅读状态，不改动任何跨平台内容字段。 */
        val MIGRATION_1_2 = object : Migration(1, 2) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL(
                    """
                    CREATE TABLE IF NOT EXISTS `idea_view_state` (
                        `logId` TEXT NOT NULL,
                        `viewedAt` TEXT NOT NULL,
                        PRIMARY KEY(`logId`),
                        FOREIGN KEY(`logId`) REFERENCES `logs`(`id`)
                            ON UPDATE NO ACTION ON DELETE CASCADE
                    )
                    """.trimIndent()
                )
                // v1 没有阅读状态，不能把所有历史 Idea 冒充为“从未查看”。
                // 以原创建时间作为迁移基线；v2 后新建/新导入的 Idea 才进入提醒队列。
                db.execSQL(
                    """
                    INSERT OR IGNORE INTO `idea_view_state` (`logId`, `viewedAt`)
                    SELECT `id`, `createdAt` FROM `logs` WHERE `category` = 'idea'
                    """.trimIndent()
                )
            }
        }

        val MIGRATION_2_3 = object : Migration(2, 3) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL(
                    """
                    CREATE TABLE IF NOT EXISTS `tasks` (
                        `id` TEXT NOT NULL,
                        `title` TEXT NOT NULL,
                        `notes` TEXT,
                        `colorTag` TEXT NOT NULL,
                        `importance` INTEGER NOT NULL,
                        `dueKind` TEXT NOT NULL,
                        `dueDate` TEXT,
                        `dueAt` TEXT,
                        `timeZone` TEXT,
                        `reminderAt` TEXT,
                        `completedAt` TEXT,
                        `createdAt` TEXT NOT NULL,
                        `updatedAt` TEXT NOT NULL,
                        PRIMARY KEY(`id`)
                    )
                    """.trimIndent()
                )
            }
        }

        /** Add indexes used by the timeline, calendar and task ordering paths. */
        val MIGRATION_3_4 = object : Migration(3, 4) {
            override fun migrate(db: SupportSQLiteDatabase) {
                db.execSQL("CREATE INDEX IF NOT EXISTS `index_logs_createdAt` ON `logs` (`createdAt`)")
                db.execSQL("CREATE INDEX IF NOT EXISTS `index_logs_recordDate` ON `logs` (`recordDate`)")
                db.execSQL(
                    "CREATE INDEX IF NOT EXISTS `index_logs_category_createdAt` " +
                        "ON `logs` (`category`, `createdAt`)"
                )
                db.execSQL(
                    "CREATE INDEX IF NOT EXISTS `index_emotions_createdAt` ON `emotions` (`createdAt`)"
                )
                db.execSQL(
                    "CREATE INDEX IF NOT EXISTS `index_emotions_recordDate` ON `emotions` (`recordDate`)"
                )
                db.execSQL("CREATE INDEX IF NOT EXISTS `index_tasks_updatedAt` ON `tasks` (`updatedAt`)")
                db.execSQL("CREATE INDEX IF NOT EXISTS `index_tasks_dueDate` ON `tasks` (`dueDate`)")
                db.execSQL("CREATE INDEX IF NOT EXISTS `index_tasks_dueAt` ON `tasks` (`dueAt`)")
            }
        }
    }
}
